aplib.net-demo/Assets/Scripts/WFC/Grid.cs

```diff
@@ -76,5 +111,119 @@ public void PlaceRoom(int x, int y, Room room)
+
+        /// <summary>
+        /// Given a cell, this method returns all directly adjacent cells, excluding diagonal neighbours.
+        /// When the cell is on an edge, it will return less than 4 neighbours.
+        /// </summary>
+        /// <param name="cell">The centre cell, who's neighbours need to be determined.</param>
+        /// <returns>All neighbouring cells, excluding diagonal neighbours.</returns>
+        /// <remarks>This does not check if the tiles in the cells are connected.</remarks>
+        /// <seealso cref="GetConnectedNeighbours"/>
+        public ICollection<Cell> Get4NeighbouringCells(Cell cell)
+        {
+            ICollection<Cell> neighbours = new Collection<Cell>();
+            if (cell.X > 0)          neighbours.Add(this[cell.X - 1, cell.Y]);
+            if (cell.X < Width - 1)  neighbours.Add(this[cell.X + 1, cell.Y]);
+            if (cell.Y > 0)          neighbours.Add(this[cell.X, cell.Y - 1]);
+            if (cell.Y < Height - 1) neighbours.Add(this[cell.X, cell.Y + 1]);
+            return neighbours;
+        }
+        
+        /// <summary>
+        /// Given a cell, this method returns all directly adjacent cells, including diagonal neighbours.
+        /// When the cell is on an edge, it will return less than 8 neighbours.
+        /// </summary>
+        /// <param name="cell">The centre cell, who's neighbours need to be determined.</param>
+        /// <returns>All neighbouring cells, including diagonal neighbours.</returns>
+        /// <remarks>This does not check if the tiles in the cells are connected.</remarks>
+        /// <seealso cref="GetConnectedNeighbours"/>
+        public ICollection<Cell> Get8NeighbouringCells(Cell cell)
+        {
+            ICollection<Cell> neighbours = new Collection<Cell>();
+            for (int i = -1; i < 2; i++)
+            {
+                if (cell.X + i < 0 || cell.X + i >= Width) continue; // No out of range
+                for (int j = -1; j < 2; j++)
+                {
+                    if (i == j && i == 0) continue; // Skip the cell itself
+                    if (cell.Y + j < 0 || cell.Y + j >= Height) continue; // No out of range
+                    neighbours.Add(this[cell.X + i, cell.Y + j]);
+                }
+            }
+            return neighbours;
+        }
+        
+        /// <summary>
+        /// Given a cell, this method determines all directly adjacent neighbours which are connected by their <see cref="Tile"/>s.
+        /// These are at most 4, as diagonal tiles cannot be connected.
+        /// </summary>
+        /// <param name="cell">The cell whose connected tiles are to be determined.</param>
+        /// <returns>The cells directly connected through their tiles</returns>
+        /// <remarks>Assumes that the cells are assigned a tile.</remarks>
+        public ICollection<Cell> GetConnectedNeighbours(Cell cell)
+        {
+            ICollection<Cell> connectedNeighbours = new Collection<Cell>();
```

Comment:
If you use a more specific class, you must also use this as the type. This is also true for the output of the function. Function inputs should be the least restrictive, while outputs should be most specific. See the google c# guide.

Reply:
The same is true for a lot of functions in this class. Please fix those too.

Reply:
https://stackoverflow.com/questions/1072614/should-i-always-return-ienumerablet-instead-of-ilistt/1924122#1924122

I would argue that when returning something, you should be as specific as possible.

If the user receives an IEnumerable, but would like to access the last item only. Then he would need to iterate (and insert 10 other complications with IEnumerables). Meanwhile, it was an IList all along, so he could have just received it as an IList and be done with it. In short, why remove details/knowledge on the returned item?

That said, why does Rider suggest changing it to an IEnumerable? Did you configure so in the `.editorconfig`?

Reply:
Now I re-read your comment

> Function inputs should be the least restrictive, while outputs should be most specific. See the google c# guide.

You agree with me here. So then, ICollection implements IEnumerable. So I am being more specific. What am I missing?

Reply:
You are using the `ICollection` interface instead of the `Collection` type, which could add more functionality.

 As per the rider suggestion, I don't know. This is not a feature of .editorconfig. I personally don't experience it either.

Reply:
When declaring `connectedNeighbours` I have to be as generic as possible, thus I use an `ICollection` interface. When declaring what the method will return, I must be as specific as possible. Since I store an `ICollection` and not just a `Collection`, I can't return specifically a `Collection` as that would require this method to assume more specifics.

This is a heavy nitpick though. But do you understand why I chose this? I don't believe `ICollection` has less functionallities than `Collection`, seeing their naming.